workflows4s-web-ui/src/main/scala/workflows4s/web/ui/subs/InstancesManager.scala

```diff
@@ -82,26 +172,130 @@ final case class InstancesManager(
+            onClick(InstancesManager.Msg.ToggleJsonState),
+          )(if showJsonState then "Hide State" else "Show State"),
+        ),
+        div(cls := "control")(
+          button(
+            cls := "button is-success is-small",
+            onClick(InstancesManager.Msg.ToggleMermaidViewer),
+          )(if showMermaidViewer then "Hide Diagram" else "Show Diagram"),
+        ),
+      ),
       if showJsonState then jsonStateViewer(instance) else div(),
     )
 
+  private def progressVisualizationView(): Html[InstancesManager.Msg] =
+    div(cls := "mt-5")(
+      h4("Workflow Progress"),
+      progressState match {
+        case InstancesManager.ProgressState.Loading =>
+          ReusableViews.loadingSpinner("Loading progress...")
+
+        case InstancesManager.ProgressState.Failed(reason) =>
+          div(cls := "notification is-warning is-light")(
+            text(s"Could not load progress: $reason"),
+          )
+
+        case InstancesManager.ProgressState.Success(progress) =>
+          div(
+            progressSummaryView(progress),
+            if showMermaidViewer then div(cls := "mt-4")(
+              text("Loading diagram..."),
+            )
+            else div(),
+          )
+
+        case InstancesManager.ProgressState.SuccessWithMermaid(progress, mermaid) =>
+          div(
+            progressSummaryView(progress),
+            if showMermaidViewer then mermaidDiagramView(mermaid) else div(),
+          )
+
+        case InstancesManager.ProgressState.Ready =>
+          div()
+      },
+    )
+
+  private def progressSummaryView(progress: ProgressResponse): Html[InstancesManager.Msg] =
+    div(cls := "box")(
+      h5("Progress Summary"),
+      pre(cls := "content is-small")(
+        code(progress.progress.spaces2),
+      ),
+    )
+
+  private def mermaidDiagramView(mermaid: String): Html[InstancesManager.Msg] =
+    div(cls := "box mt-4")(
+      h5("🎨 Workflow Diagram"),
+
+      // Action buttons
+      div(cls := "field is-grouped mb-4")(
+        div(cls := "control")(
+          a(
+            cls    := "button is-success is-small",
+            href   := createMermaidLiveUrl(mermaid),
+            target := "_blank",
+          )("🔗 View in Mermaid Live"),
+        ),
+      ),
+      renderedMermaidSvg match {
+        case Some(svg) =>
+          div(cls := "has-text-centered p-4")().innerHtml(svg)
+        case None      =>
+          div(cls := "notification is-info is-light has-text-centered")(
+            text("🔄 Rendering diagram..."),
+          )
+      },
+
+      // Code view
+      details(cls := "mt-4")(
+        summary(cls := "button is-small is-light")("📋 View Mermaid Code"),
+        pre(cls := "mt-2")(
+          code(mermaid),
+        ),
+      ),
+    )
+  private def createMermaidLiveUrl(mermaidCode: String): String               = {
+    val encoded = java.net.URLEncoder.encode(mermaidCode, "UTF-8")
+    s"https://mermaid.live/edit#code:$encoded"
```

Comment:
I think there is a bit of compelxity into generating those urls correctly. This is already implemented inside `MermaidFlowchart` (or somehwere around). Lets reuse this.